scripts/test_usb_scan.py

```diff
@@ -0,0 +1,393 @@
+#
+# unittest for usb_scan.py
+
+from collections import Mapping, Container, Iterable
+import mock
+from nose.tools import nottest
+import os
+import shutil
+import sys
+import tempfile
+import unittest
+
+sys.modules["xcp"] = mock.Mock()
+sys.modules["xcp.logger"] = mock.Mock()
+sys.modules["pyudev"] = mock.Mock()
+
+log_file = None
+
+
+class MocDeviceAttrs(Mapping):
+    def __init__(self, device):
+        self.d = device.get_attr()
+
+    def __iter__(self):
+        for name in self.d:
+            yield name
+
+    def __len__(self):
+        return len(self.d)
+
+    def __getitem__(self, name):
+        return self.d.get(name)
+
+
+class MocDevice(Mapping):
+
+    def __init__(self, d):
+        self.d = d
+
+    @property
+    def sys_name(self):
+        return self.d.get("name")
+
+    def get_prop(self):
+        return self.d.get("props")
+
+    def get_attr(self):
+        return self.d.get("attrs")
+
+    @property
+    def attributes(self):
+        return MocDeviceAttrs(self)
+
+    def __iter__(self):
+        for name in self.get_prop():
+            yield name
+
+    def __len__(self):
+        return len(self.get_prop())
+
+    def __getitem__(self, name):
+        return self.get_prop().get(name)
+
+
+class MocEnumerator(object):
+
+    def __init__(self, ds):
+        self.ds = ds
+
+    def __iter__(self):
+        for d in self.ds:
+            yield MocDevice(d)
+
+
+class MocContext(object):
+
+    def __init__(self, devices, interfaces):
+        self.devices = devices
+        self.interfaces = interfaces
+
+    def list_devices(self, **kwargs):
+        if "usb" == kwargs.pop("subsystem"):
+            dev_type = kwargs.pop("DEVTYPE")
+            if "usb_device" == dev_type:
+                return MocEnumerator(self.devices)
+            elif "usb_interface" == dev_type:
+                return MocEnumerator(self.interfaces)
+        return MocEnumerator([])
+
+
+def mock_setup(mod, devices, interfaces, path):
+    mod.log.error = test_log
+    mod.log.debug = test_log
+    mod.Policy._PATH = path
+    mod.pyudev.Context = mock.Mock(return_value=MocContext(
+        devices, interfaces))
+
+
+@nottest
+def test_log(m):
+    global log_file
+    with open(log_file, "a+") as f:
+        f.write("{}: {}\n".format(sys.argv[0], m))
+
+
+class TestUsbScan(unittest.TestCase):
+
+    def setUp(self):
+        global log_file
+        try:
+            self.work_dir = tempfile.mkdtemp(prefix="test_usb_scan")
+            log_file = os.path.join(self.work_dir, "user.log")
+        except:
+            raise
+
+    def tearDown(self):
+        shutil.rmtree(self.work_dir, ignore_errors=True)
+
+    @nottest
+    def test_usb_common(self, moc_devices, moc_interfaces, moc_results,
+                        path="./scripts/usb-policy.conf"):
+        import usb_scan
+        mock_setup(usb_scan, moc_devices, moc_interfaces, path)
+
+        devices, interfaces = usb_scan.get_usb_info()
+
+        # debug info
```

Comment:
We should not copy paste the code, instead we should refactor `usb_scan.py` to be better suited for testing.
I think we need to extract https://github.com/xapi-project/xen-api/blob/3bb747104ee0399c3d0e3c747b1d8ff95669b269/scripts/usb_scan.py#L627-L637 into a function that takes `devices` and `interfaces` as parameters and returns the object that we will serialise to json. Then this test should call that function with the mocked devices and interfaces

Reply:
OK. This will change the usb_scan.py,
I have to run more test next week.

Reply:
Yes, call it something like `make_pusbs_list(interfaces, devices)` and the unit test that function